runtime/gc_vlhgc/MemorySubSpaceTarok.cpp

```diff
@@ -952,12 +954,233 @@ void
 MM_MemorySubSpaceTarok::checkResize(MM_EnvironmentBase *env, MM_AllocateDescription *allocDescription, bool _systemGC)
 {
 	UDATA oldVMState = env->pushVMstate(OMRVMSTATE_GC_CHECK_RESIZE);
-	if (!timeForHeapContract(env, allocDescription, _systemGC)) {
-		timeForHeapExpand(env, allocDescription);
+
+	IDATA heapSizeChange = calculateHeapSizeChange(env, allocDescription, _systemGC);
+
+	if (heapSizeChange < 0) {
+		_contractionSize = (UDATA)(heapSizeChange * -1);
+		_expansionSize = 0;
+	} else if (heapSizeChange > 0) {
+		_contractionSize = 0;
+		_expansionSize = (UDATA)heapSizeChange;
+	} else {
+		_contractionSize = 0;
+		_expansionSize = 0;
 	}
+	/* Certain edge cases will trigger repetitive Global Gc's (which need to be treated differently), and this is tracked by observing pgc count */
+	_previouslyObservedPGCCount = ((MM_EnvironmentVLHGC *)env)->_heapSizingData.pgcCountSinceGMPEnd;
+
 	env->popVMstate(oldVMState);
 }
 
+/**
+ * Calculate by how many bytes we should change the current heap size. 
+ * @return positive number of bytes represents how many bytes the heap should expand. 
+ * 		   If heap should contract, a negative number representing how many bytes the heap should contract is returned.
+ */
+IDATA
+MM_MemorySubSpaceTarok::calculateHeapSizeChange(MM_EnvironmentBase *env, MM_AllocateDescription *allocDescription, bool _systemGC) 
+{
+	IDATA sizeChange = 0;
+
+	/* Heap sizing is driven by "hybrid Heap overhead". This is a blended value, which combines free memory in "tenure" along with observed GC overhead.
+	 * If the hybrid score is too high (can be acheived by low free memory %, high gc %, or a hybrid of the 2), then the heap calculates an expansion value.
+	 * If the hybrid score is too low (can be acheived with high free memory %, low gc %, or hybrid of the 2), then the heap calculates a contraction value.
+	 * 
+	 * Notes: - If gc % is very low (suggesting contraction), and memory % is also very low (suggesting expansion), the hybrid heap score will likely suggest expansion to avoid OOM error
+	 *        - If the hybrid value falls within the acceptable thresholds, the heap will not resize
+	 */
+	double hybridHeapScore = calculateCurrentHybridHeapOverhead(env);
+	
+	/* Based on the hybrid overhead of gc cpu, and free memory, decide if heap should expand or contract */
+	if (hybridHeapScore > (double)_extensions->heapExpansionGCTimeThreshold) {
+		/* Try to expand the heap */
+		sizeChange = (IDATA)calculateExpansionSize(env, allocDescription, _systemGC);
+	} else if (hybridHeapScore < (double)_extensions->heapContractionGCTimeThreshold) {
+		/* Try to contract the heap */
+		sizeChange = calculateContractionSize(env, allocDescription, _systemGC, true);
+	}
+
+	if (0 == sizeChange && hybridHeapScore >= (double)_extensions->heapContractionGCTimeThreshold) {
```

Comment:
As per coding standards, we put extra oval brackets and don't rely on operator priorities:

((0 == sizeChange) && (hybridHeapScore >= (double)_extensions->heapContractionGCTimeThreshold))

Please go through all the changes and add them when necessary. For this pass I'll ignore the rest of missing ones.